ros/src/waypoint_updater/waypoint_updater.py

```diff
@@ -60,22 +81,53 @@ def obstacle_cb(self, msg):
         # We will implement it later
         pass
 
-    def get_waypoint_velocity(self, waypoint):
+    def find_next_waypoint(self):
+        min_dist = 1e10
+        min_idx = -1
+        if self.ego and self.world:
+            # Find the closest waypoint
+            for i in range(len(self.world.waypoints)):
+                wp = self.world.waypoints[i]
+                dl = self.dist(self.ego.pose.position, wp.pose.pose.position)
+                if dl < min_dist:
+                    min_dist = dl
+                    min_idx = i
+
+            # Check if we are behind or past the closest waypoint
+            wp = self.world.waypoints[min_idx]
+            pos = self.ego.pose.position
+            pos.x += self.ego.pose.orientation.x
+            pos.y += self.ego.pose.orientation.y
```

Comment:
I don't understand this. how is adding orientation x,y determining if we are ahead or behind the waypoint?
Also, the orientation (x,y,z,w) is in Quaternion (whatever that means). I tried turning car around, printing all values and x,y remain 0. Need to convert this somehow to different method.

Reply:
Good catch! Indeed the orientation is a Quaternion. I believe `tf` has function for creating a rotation matrix from it.

Reply:
what is tf?

Reply:
Totally missed that one! Thanks @orenmeiri for catching it. `tf` seems to be a package which can help with transforming Quaternion's to something more understandable :smile: I will see if I can get an orientation vector in Cartesian coordinates and use that instead.

Reply:
The logic is as follows: move some units along the current orientation and if that makes us closer to the closest waypoint, then we are behind it and it is the "next waypoint". Otherwise, it is behind us and it was the "last waypoint" and the following one is the "next waypoint". There are logs for debugging it at line 53.